internal/convey/convey.go

```diff
@@ -0,0 +1,115 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package convey
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/xmidt-org/xmidt-agent/internal/net"
+)
+
+type Option interface {
+	apply(*ConveyHeaderProvider) error
+}
+
+type optionFunc func(*ConveyHeaderProvider) error
+
+func (f optionFunc) apply(c *ConveyHeaderProvider) error {
+	return f(c)
+}
+
+const (
+	HeaderName = "X-Webpa-Convey"
+)
+
+const (
+	Firmware                   = "fw-name"
+	Hardware                   = "hw-model"
+	Manufacturer               = "hw-manufacturer"
+	SerialNumber               = "hw-serial-number"
+	LastRebootReason           = "hw-last-reboot-reason"
+	Protocol                   = "webpa-protocol"
+	BootTime                   = "boot-time"
+	BootTimeRetryDelay         = "boot-time-retry-wait"
+	InterfaceUsed       string = "webpa-interface-used"
+	InterfacesAvailable        = "interfaces-available"
+)
+
+type ConveyHeaderProvider struct {
+	networkService     net.NetworkServicer
+	fields             []string
+	firmware           string
+	hardware           string
+	manufacturer       string
+	serialNumber       string
+	lastRebootReason   string
+	protocol           string
+	bootTime           string
+	bootTimeRetryDelay string
+}
+
+func New(opts ...Option) (*ConveyHeaderProvider, error) {
+	conveyHeaderProvider := &ConveyHeaderProvider{}
+
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(conveyHeaderProvider); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return conveyHeaderProvider, nil
+}
+
+func (c *ConveyHeaderProvider) GetConveyHeader() map[string]interface{} {
```

Comment:
Instead of getting the header, I'd just make the ConveyHeaderProvider have a few decorators (one for HTTP and one for WRP)

Reply:
I want to keep that function so I can use it for both decorators